2016/proposals/pravendra-singh.md

```diff
@@ -0,0 +1,220 @@
+
+## Technical Details
+
+The application will be powered by Django, as we also want it to work with AMY. It will use Plotly.js for visualizing different response types, which is a open source [project](https://plot.ly/javascript/open-source-announcement/).
+
+Going by the "*Object-oriented analysis and design*" approach for system design, we can categorize the application in the following sections.
+
+###### 1. Visual representation of the survey questions
+
+Both the type of surveys have multiple questions of varying classes. Ideally, a good visualizer should respect the inherent attributes of the dataset. That's why the application will employ suitable visualization for different types of questions.
+
+According to this [explanation](https://github.com/numfocus/gsoc/issues/77#issuecomment-192500549), here is a list of suitable charts.
+
+| Question Type  | Suitable Chart Type | Sample Question |
+| ------------- | ------------- | ------------- |
+| Multiple choice - Single answer  | Bar and Pie chart  | When are you taking this survey? |
+| Multiple choice - Multiple answer  | Chord diagram  |  What is your domain of research/study? |
+| Yes-No-Maybe type  | Horizontal Stacked Bar chart  |  How would you describe your ability to do the following tasks? |
+| Matrix type  | Heatmap Matrix chart  | How would you rate your motivation to learn about these topics? |
+| Text as an input  | Text-Cloud OR Table OR Semantic Analysis | In a few words, what is your most important reason for attending this workshop? |
+
+###### 2. Response-Cache service
+
+As we are already aware of the order of total responses we get from our attendees, having a "*Data Cache Layer*" is an intelligent solution. Rate Limitations exposed by SurveyMonkey API is another reason behind implementing a cache service.
+
+The *Response-Cache Service* will be responsible for serving cached responses, and hence plots. It will leverage the *offline* mode of Plotly so that the caching service is only dependent on the API responses and idempotent in terms of generated plots.
+
+The service will use *Heartbeat Mechanism* to invalidate cache content with respect to the API response periodically. Workshop identifiers (ID) will be used to collect and invalidate responses for a particular event.
+
+**Interaction between SurveyMonkey API and Response-Cache Service**
+```
+           +-------------------+        
+           |    SurveyMonkey   |        
+           |         API       |        
+           +--^-------------^--+        
+              |  Heartbeat  |           
+           +--v-------------v--+        
+           |   Response Cache  |        
+           |      Service      |        
+           +-------------------+
+```
+
+This service consists of two elements "Cache Storage" [CS] and "Permanent Storage" [PS], using "Redis" and "RethinkDB" respectively.
```

Comment:
Hi,

> Charting libraries, be it plotly.js or any other, accepts the input data in the form of JSON objects, so I will prefer storing the response data (that I want to plot for a survey) as JSON.

And you're sure the JSON from Survey Monkey won't need to be reformatted or something before it's accepted by plotly.js?

> But some portion of data is not suitable as JSON, or storing it in this format will be bad as we will have to format it over and over. For example, we need to match/select the responses according to the "workshop_id" or "workshop's name", and there is no need of JSON for it.

Yes, filtering is a quite important feature of this system.